File: server-csharp/AttackUtils.cs
using SpacetimeDB;
using System;

public static partial class Module
{
    // Utility class for attack-related helper functions
    public static class AttackUtils
    {
        public static uint GetParameterU(ReducerContext ctx, PlayerScheduledAttack attack)
        {
            var playerId = attack.player_id;
            var attackType = attack.attack_type;
            
            var playerOpt = ctx.Db.player.player_id.Find(playerId);
            if (playerOpt == null)
            {
                throw new Exception($"GetParameterU: Player {playerId} not found");
            }

            var player = playerOpt.Value;

            var entityOpt = ctx.Db.entity.entity_id.Find(player.entity_id);
            if (entityOpt == null)
            {
                throw new Exception($"GetParameterU: Entity {player.entity_id} not found for player {playerId}");
            }

            var pEntity = entityOpt.Value;
            
            switch (attackType)
            {
                case AttackType.Sword:
                {
                    if (attack.parameter_u == 0)
                    {
                        return 1;
                    }
                    else
                    {
                        return 0;
                    }
                }
                case AttackType.Knives:
                {
                    //Random angle on a circle
                    var random = ctx.Rng;
                    var angle = random.NextDouble() * 360.0;
                    return (uint)angle;
                }
                case AttackType.Shield:
                {
                    var random = ctx.Rng;
                    var angle = random.NextDouble() * 360.0;
                    return (uint)angle;
                }
                case AttackType.ThrowingShield:
                {
                    // For throwing shield, parameter_u stores the bounce count
                    return 0; // Start with 0 bounces
                }
                case AttackType.EnergyOrb:
                {
                    var random = ctx.Rng;
                    var angle = random.NextDouble() * 360.0;
                    return (uint)angle; // Similar to Shield's implementation
                }
                case AttackType.MagicDagger:
                {
                    // For magic dagger, parameter_u determines if it's returning (0=going out, 1=returning)
                    return 0; // Start in outgoing state
                }
                default:
                {
                    return 0;
                }
            }
        }   

        // Determine the direction of the attack based on attack type and other factors
        public static DbVector2 DetermineAttackDirection(ReducerContext ctx, uint playerId, AttackType attackType, uint idWithinBurst, uint parameterU, int parameterI)
        {
            // Get the player
            var playerOpt = ctx.Db.player.player_id.Find(playerId);
            if (playerOpt == null)
            {
                throw new Exception($"DetermineAttackDirection: Player {playerId} not found");
            }

            var player = playerOpt.Value;
            
            // Get the entity
            var entityOpt = ctx.Db.entity.entity_id.Find(player.entity_id);
            if (entityOpt == null)
            {
                throw new Exception($"DetermineAttackDirection: Entity {player.entity_id} not found for player {playerId}");
            }

            var entity = entityOpt.Value;

            var attackData = FindAttackDataByType(ctx, attackType);
            if (attackData == null)
            {
                throw new Exception($"DetermineAttackDirection: Attack data not found for type {attackType}");
            }

            // Handle different attack types
            switch (attackType)
            {
                case AttackType.Sword:
                {
                    // Sword attacks now target the nearest enemy like wands for better kiting
                    Entity? nearestEnemy = FindNearestEnemy(ctx, entity);
                    if (nearestEnemy != null)
                    {
                        var enemyActual = nearestEnemy.Value;

                        // Calculate direction vector to the enemy
                        var dx = enemyActual.position.x - entity.position.x;
                        var dy = enemyActual.position.y - entity.position.y;
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            return new DbVector2((float)(dx / length), (float)(dy / length));
                        }
                    }
                    
                    // If no enemies or calculation issue, fall back to alternating left/right pattern
                    var countParam = parameterU + idWithinBurst;
                    if (countParam % 2 == 0)
                    {
                        return new DbVector2(1, 0);
                    }
                    else
                    {
                        return new DbVector2(-1, 0);
                    }
                }
                case AttackType.Wand:
                {
                    // Wands shoot at the nearest enemy
                    Entity? nearestEnemy = FindNearestEnemy(ctx, entity);
                    if (nearestEnemy != null)
                    {
                        var enemyActual = nearestEnemy.Value;

                        // Calculate direction vector to the enemy
                        var dx = enemyActual.position.x - entity.position.x;
                        var dy = enemyActual.position.y - entity.position.y;
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            return new DbVector2((float)(dx / length), (float)(dy / length));
                        }
                    }
                    
                    // If no enemies or calculation issue, use player's direction
                    return GetNormalizedDirection(entity.direction);
                }
                case AttackType.Knives:
                {
                    //Knives attack in a circle around the player starting at the angle specified in the parameter_u
                    //The angle is in degrees, so we need to convert it to radians
                    var startAngle = (double)parameterU * Math.PI / 180.0;                       
                    var angleStep = 360.0 / (double)attackData.Value.projectiles;
                    var attackAngle = startAngle + (angleStep * (double)idWithinBurst);
                    return new DbVector2((float)Math.Cos(attackAngle), (float)Math.Sin(attackAngle));   
                }
                case AttackType.Shield:
                {
                    // Shield attacks have rotation motion
                    return new DbVector2(0, 0);
                }
                case AttackType.Football:
                {
                    // Football attacks now target the nearest enemy like wands
                    Entity? nearestEnemy = FindNearestEnemy(ctx, entity);
                    if (nearestEnemy != null)
                    {
                        var enemyActual = nearestEnemy.Value;

                        // Calculate direction vector to the enemy
                        var dx = enemyActual.position.x - entity.position.x;
                        var dy = enemyActual.position.y - entity.position.y;
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            return new DbVector2((float)(dx / length), (float)(dy / length));
                        }
                    }
                    
                    // If no enemies or calculation issue, use player's direction
                    return GetNormalizedDirection(entity.direction);
                }
                case AttackType.Cards:
                {
                    // Find the nearest enemy for card targeting
                    Entity? nearestEnemy = FindNearestEnemy(ctx, entity);
                    if (nearestEnemy != null)
                    {
                        var enemyActual = nearestEnemy.Value;
                        
                        // Calculate base direction vector to the enemy
                        var dx = enemyActual.position.x - entity.position.x;
                        var dy = enemyActual.position.y - entity.position.y;
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            // Normalize base direction
                            var baseDirX = dx / length;
                            var baseDirY = dy / length;
                            
                            // For multiple cards, spread them in a fan pattern toward the enemy
                            // The spread angle depends on the number of projectiles
                            double fanAngleRange = 45.0; // degrees total spread
                            
                            // Calculate the angle for this specific card within the fan
                            double fanAngle;
                            if (attackData.Value.projectiles > 1)
                            {
                                // Calculate angle offset for this projectile in the fan
                                fanAngle = -fanAngleRange / 2.0 + (fanAngleRange * idWithinBurst / (attackData.Value.projectiles - 1));
                            }
                            else
                            {
                                fanAngle = 0;
                            }
                            
                            // Convert angle to radians
                            var fanAngleRad = fanAngle * Math.PI / 180.0;
                            
                            // Rotate the base vector by the fan angle
                            var rotatedX = baseDirX * Math.Cos(fanAngleRad) - baseDirY * Math.Sin(fanAngleRad);
                            var rotatedY = baseDirX * Math.Sin(fanAngleRad) + baseDirY * Math.Cos(fanAngleRad);
                            
                            return new DbVector2((float)rotatedX, (float)rotatedY);
                        }
                    }
                    
                    // If no enemies found, fall back to the original pattern
                    var startAngle = (double)parameterU * Math.PI / 180.0;                       
                    var angleStep = 360.0 / (double)attackData.Value.projectiles;
                    var attackAngle = startAngle + (angleStep * (double)idWithinBurst);
                    return new DbVector2((float)Math.Cos(attackAngle), (float)Math.Sin(attackAngle));
                }
                case AttackType.Dumbbell:
                {
                    // Dumbbells start with a strong upward motion
                    var random = ctx.Rng;
                    var yOffset = -4.0f; // Stronger upward initial velocity
                    var xOffset = (random.NextDouble() - 0.5) * 1.0f; // Reduced horizontal spread
                    return new DbVector2((float)xOffset, yOffset).Normalize(); // Normalized direction vector with strong upward motion
                }
                case AttackType.Garlic:
                {
                    // Garlic stays with player, no movement needed
                    return new DbVector2(0, 0);
                }
                case AttackType.Shuriken:
                {
                    // Shurikens throw in a spread pattern toward the nearest enemy with a spinning trajectory
                    Entity? nearestEnemy = FindNearestEnemy(ctx, entity);
                    if (nearestEnemy != null)
                    {
                        var enemyActual = nearestEnemy.Value;
                        
                        // Calculate base direction vector to the enemy
                        var dx = enemyActual.position.x - entity.position.x;
                        var dy = enemyActual.position.y - entity.position.y;
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            // Normalize base direction
                            var baseDirX = dx / length;
                            var baseDirY = dy / length;
                            
                            // For multiple shurikens, spread them in a narrow fan pattern
                            double fanAngleRange = 30.0; // degrees total spread (narrower than cards)
                            
                            // Calculate the angle for this specific shuriken within the fan
                            double fanAngle;
                            if (attackData.Value.projectiles > 1)
                            {
                                // Calculate angle offset for this projectile in the fan
                                fanAngle = -fanAngleRange / 2.0 + (fanAngleRange * idWithinBurst / (attackData.Value.projectiles - 1));
                            }
                            else
                            {
                                fanAngle = 0;
                            }
                            
                            // Convert angle to radians
                            var fanAngleRad = fanAngle * Math.PI / 180.0;
                            
                            // Rotate the base vector by the fan angle
                            var rotatedX = baseDirX * Math.Cos(fanAngleRad) - baseDirY * Math.Sin(fanAngleRad);
                            var rotatedY = baseDirX * Math.Sin(fanAngleRad) + baseDirY * Math.Cos(fanAngleRad);
                            
                            return new DbVector2((float)rotatedX, (float)rotatedY);
                        }
                    }
                    
                    // If no enemies found, fall back to player direction
                    return GetNormalizedDirection(entity.direction);
                }
                case AttackType.FireSword:
                {
                    // Fire sword attacks in the direction the player is facing with a fiery trail
                    return GetNormalizedDirection(entity.direction);
                }
                case AttackType.HolyHammer:
                {
                    // Holy hammer targets the nearest enemy with a strong directional attack
                    Entity? nearestEnemy = FindNearestEnemy(ctx, entity);
                    if (nearestEnemy != null)
                    {
                        var enemyActual = nearestEnemy.Value;
                        
                        // Calculate direction vector to the enemy
                        var dx = enemyActual.position.x - entity.position.x;
                        var dy = enemyActual.position.y - entity.position.y;
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            return new DbVector2((float)(dx / length), (float)(dy / length));
                        }
                    }
                    
                    // If no enemies or calculation issue, use player's direction
                    return GetNormalizedDirection(entity.direction);
                }
                case AttackType.MagicDagger:
                {
                    // For outgoing daggers, aim at nearest enemy
                    if (parameterU == 0) // Outgoing state
                    {
                        Entity? nearestEnemy = FindNearestEnemy(ctx, entity);
                        if (nearestEnemy != null)
                        {
                            var enemyActual = nearestEnemy.Value;
                            
                            // Calculate base direction vector to the enemy
                            var dx = enemyActual.position.x - entity.position.x;
                            var dy = enemyActual.position.y - entity.position.y;
                            
                            // Normalize the direction
                            var length = Math.Sqrt(dx * dx + dy * dy);
                            if (length > 0)
                            {
                                // Normalize base direction
                                var baseDirX = dx / length;
                                var baseDirY = dy / length;
                                
                                // For multiple daggers, spread them slightly
                                double fanAngleRange = 20.0; // degrees total spread
                                
                                // Calculate the angle for this specific dagger within the fan
                                double fanAngle;
                                if (attackData.Value.projectiles > 1)
                                {
                                    // Calculate angle offset for this projectile in the fan
                                    fanAngle = -fanAngleRange / 2.0 + (fanAngleRange * idWithinBurst / (attackData.Value.projectiles - 1));
                                }
                                else
                                {
                                    fanAngle = 0;
                                }
                                
                                // Convert angle to radians
                                var fanAngleRad = fanAngle * Math.PI / 180.0;
                                
                                // Rotate the base vector by the fan angle
                                var rotatedX = baseDirX * Math.Cos(fanAngleRad) - baseDirY * Math.Sin(fanAngleRad);
                                var rotatedY = baseDirX * Math.Sin(fanAngleRad) + baseDirY * Math.Cos(fanAngleRad);
                                
                                return new DbVector2((float)rotatedX, (float)rotatedY);
                            }
                        }
                        
                        // If no enemies found, fall back to player direction
                        return GetNormalizedDirection(entity.direction);
                    }
                    else // Returning state (parameterU == 1)
                    {
                        // Calculate direction back to the player
                        var dx = entity.position.x - parameterI;  // parameterI stores the entity_id of the player
                        var dy = entity.position.y - parameterI;  // This will be updated in the attack handler
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            return new DbVector2((float)(-dx / length), (float)(-dy / length));
                        }
                        
                        // Fallback
                        return new DbVector2(-1, 0);
                    }
                }
                case AttackType.ThrowingShield:
                {
                    // Throwing shield initially targets nearest enemy, then bounces to others
                    Entity? targetEnemy = FindNearestEnemy(ctx, entity);
                    if (targetEnemy != null)
                    {
                        var enemyActual = targetEnemy.Value;
                        
                        // Calculate direction vector to the enemy
                        var dx = enemyActual.position.x - entity.position.x;
                        var dy = enemyActual.position.y - entity.position.y;
                        
                        // Normalize the direction
                        var length = Math.Sqrt(dx * dx + dy * dy);
                        if (length > 0)
                        {
                            return new DbVector2((float)(dx / length), (float)(dy / length));
                        }
                    }
                    
                    // If no enemies or calculation issue, use player's direction
                    return GetNormalizedDirection(entity.direction);
                }
                case AttackType.EnergyOrb:
                {
                    // Energy orbs orbit the player like shields but with varied positions
                    // This will be handled similar to Shield attack but with specific logic in ProcessAttackMovements
                    return new DbVector2(0, 0);
                }
                default:
                {
                    throw new Exception($"DetermineAttackDirection: Unknown attack type {attackType}");
                }
            }
        }

        // Find the nearest enemy to a player entity
        public static Entity? FindNearestEnemy(ReducerContext ctx, Entity playerEntity)
        {
            Entity? nearestEnemy = null;
            float nearestDistanceSquared = float.MaxValue;

            // Iterate through all monsters in the game
            foreach (var monster in ctx.Db.monsters.Iter())
            {
                // Get the monster's entity
                var monsterEntityOpt = ctx.Db.entity.entity_id.Find(monster.entity_id);
                if (monsterEntityOpt == null)
                {
                    continue; // Skip if entity not found
                }

                var monsterEntity = monsterEntityOpt.Value;

                // Calculate squared distance (more efficient than using square root)
                var dx = monsterEntity.position.x - playerEntity.position.x;
                var dy = monsterEntity.position.y - playerEntity.position.y;
                var distanceSquared = dx * dx + dy * dy;

                // If this monster is closer than the current nearest, update nearest
                if (distanceSquared < nearestDistanceSquared)
                {
                    nearestDistanceSquared = distanceSquared;
                    nearestEnemy = monsterEntity;
                }
            }

            return nearestEnemy;
        }

        // Helper to normalize a direction vector, with default fallback
        private static DbVector2 GetNormalizedDirection(DbVector2 direction)
        {
            // If the entity has no direction (not moving), use a default direction
            if (direction.x == 0 && direction.y == 0)
            {
                // Default to right direction if no movement direction
                return new DbVector2(1, 0);
            }
            else
            {
                // Normalize the direction vector
                var length = Math.Sqrt(direction.x * direction.x + direction.y * direction.y);
                if (length > 0)
                {
                    return new DbVector2((float)(direction.x / length), (float)(direction.y / length));
                }
                else
                {
                    return new DbVector2(1, 0); // Fallback
                }
            }
        }
    }
}

File: server-rust/src/attack_utils.rs
use spacetimedb::{ReducerContext, Table, rand::{Rng, RngCore}};
use crate::{DbVector2, AttackType, PlayerScheduledAttack, monsters_def::{monsters, monsters_boid}};
use std::f64::consts::PI;

// Thunder Horn attack configuration constants
pub const THUNDER_HORN_TARGET_RADIUS: f32 = 700.0;  // Maximum range for Thunder Horn targeting

// Get parameter_u for an attack
pub fn get_parameter_u(ctx: &ReducerContext, attack: &PlayerScheduledAttack) -> u32 {
    let attack_type = &attack.attack_type;

    match attack_type {
        AttackType::Sword => {
            if attack.parameter_u == 0 {
                1
            } else {
                0
            }
        }
        AttackType::Knives => {
            // Calculate consistent rotation based on attack count
            // Each attack should rotate by the angle between knives
            let angle_between_knives = 360.0 / attack.projectiles as f64;
            let rotation_offset = (attack.attack_count as f64 * angle_between_knives / 12.0) % 360.0;
            rotation_offset as u32
        }
        AttackType::Shield => {
            let mut rng = ctx.rng();
            let angle = rng.gen_range(0.0..360.0);
            angle as u32
        }
        AttackType::ThunderHorn => {
            // Store the target position in parameter_u and parameter_i
            // We'll pack the position coordinates into these parameters
            0 // Position will be determined in determine_attack_direction
        }
        AttackType::AngelStaff => {
            // Angel Staff doesn't need special parameters since it's a simple area effect
            0
        }
        AttackType::Football | AttackType::Cards | AttackType::Dumbbell => {
            // For now, use a simple alternating parameter like Sword
            if attack.parameter_u == 0 {
                1
            } else {
                0
            }
        }
        AttackType::Volleyball => {
            // For Volleyball, parameter_u stores the bounce count
            if attack.parameter_u == 0 {
                1
            } else {
                0
            }
        }
        AttackType::Garlic => {
            // Garlic has no direction, so no parameter needed
            0
        }
        _ => 0,
    }
}

// Determine the direction of the attack based on attack type and other factors
pub fn determine_attack_direction(
    ctx: &ReducerContext,
    player_id: u32,
    attack_type: &AttackType,
    id_within_burst: u32,
    parameter_u: u32,
    _parameter_i: i32,
    num_projectiles: u32,
) -> DbVector2 {
    // Get player entity from collision cache
    let cache = crate::monsters_def::get_collision_cache();
    let player_cache_idx = match cache.player.player_id_to_cache_index.get(&player_id) {
        Some(&idx) => idx as usize,
        None => {
            log::error!("Player {} not found in collision cache for attack direction", player_id);
            return DbVector2::new(1.0, 0.0); // Fallback direction
        }
    };
    let player_pos = DbVector2::new(cache.player.pos_x_player[player_cache_idx], cache.player.pos_y_player[player_cache_idx]);
    
    // Since dir_x and dir_y don't exist in the cache, we'll use a fallback direction
    // or calculate direction based on player's last movement or target
    let player_dir = DbVector2::new(1.0, 0.0); // Default direction (right)

    match attack_type {
        AttackType::Sword | AttackType::Joint => {
            // Sword and Joint attacks target the nearest enemy for better kiting
            if let Some(nearest_enemy_pos) = find_nearest_enemy(ctx, player_pos, player_id) {
                let dir = (nearest_enemy_pos - player_pos).normalize();
                if dir.length_sq() > 0.0 {
                    return dir;
                }
            }
            
            // If no enemies, fall back to alternating left/right pattern
            let count_param = parameter_u + id_within_burst;
            if count_param % 2 == 0 {
                DbVector2::new(1.0, 0.0)
            } else {
                DbVector2::new(-1.0, 0.0)
            }
        }
        AttackType::Wand | AttackType::Football => {
            // Wands and Footballs shoot at the nearest enemy
            if let Some(nearest_enemy_pos) = find_nearest_enemy(ctx, player_pos, player_id) {
                let dir = (nearest_enemy_pos - player_pos).normalize();
                if dir.length_sq() > 0.0 {
                    return dir;
                }
            }
            
            // If no enemies or calculation issue, use player's direction
            get_normalized_direction(player_dir)
        }
        AttackType::Knives => {
            // Knives attack in a circle around the player starting at the angle specified in parameter_u
            let start_angle = parameter_u as f64 * PI / 180.0;
            let angle_step = 360.0 / num_projectiles as f64;
            let attack_angle = start_angle + (angle_step * id_within_burst as f64);
            DbVector2::new(attack_angle.cos() as f32, attack_angle.sin() as f32)
        }
        AttackType::Cards => {
            // Find the nearest enemy for card targeting
            if let Some(nearest_enemy_pos) = find_nearest_enemy(ctx, player_pos, player_id) {
                let base_dir = (nearest_enemy_pos - player_pos).normalize();
                if base_dir.length_sq() > 0.0 {
                    // For multiple cards, spread them in a fan pattern toward the enemy
                    let fan_angle_range = 45.0; // degrees total spread
                    
                    let fan_angle = if num_projectiles > 1 {
                        -fan_angle_range / 2.0 + (fan_angle_range * id_within_burst as f64 / (num_projectiles - 1) as f64)
                    } else {
                        0.0
                    };
                    
                    let fan_angle_rad = fan_angle * PI / 180.0;
                    
                    // Rotate the base vector by the fan angle
                    let rotated_x = base_dir.x as f64 * fan_angle_rad.cos() - base_dir.y as f64 * fan_angle_rad.sin();
                    let rotated_y = base_dir.x as f64 * fan_angle_rad.sin() + base_dir.y as f64 * fan_angle_rad.cos();
                    
                    return DbVector2::new(rotated_x as f32, rotated_y as f32);
                }
            }
            
            // If no enemies found, fall back to the circular pattern like knives
            let start_angle = parameter_u as f64 * PI / 180.0;
            let angle_step = 360.0 / num_projectiles as f64;
            let attack_angle = start_angle + (angle_step * id_within_burst as f64);
            DbVector2::new(attack_angle.cos() as f32, attack_angle.sin() as f32)
        }
        AttackType::Dumbbell => {
            // Dumbbells start with a strong upward motion
            let mut rng = ctx.rng();
            let y_offset = -4.0;
            let x_offset = (rng.gen::<f64>() - 0.5) * 1.0; // gen::<f64>() returns num in [0.0, 1.0)
            DbVector2::new(x_offset as f32, y_offset as f32).normalize()
        }
        AttackType::Shield | AttackType::Garlic => {
            // These attacks have special movement, not a fixed direction
            DbVector2::new(0.0, 0.0)
        }
        AttackType::ThunderHorn => {
            // Direction is not used for placement, but can be used for graphics
            if let Some(nearest_enemy_pos) = find_nearest_enemy(ctx, player_pos, player_id) {
                let dir = (nearest_enemy_pos - player_pos).normalize();
                if dir.length_sq() > 0.0 {
                    return dir;
                }
            }
            get_normalized_direction(player_dir)
        }
        AttackType::AngelStaff => {
            // No direction needed for area effect
            DbVector2::new(0.0, 0.0)
        }
        AttackType::Volleyball => {
            // Volleyballs shoot at the nearest enemy
            if let Some(nearest_enemy_pos) = find_nearest_enemy(ctx, player_pos, player_id) {
                let dir = (nearest_enemy_pos - player_pos).normalize();
                if dir.length_sq() > 0.0 {
                    return dir;
                }
            }
            // If no enemies or calculation issue, use player's direction
            get_normalized_direction(player_dir)
        }
    }
}

// Find the nearest enemy to a given position
pub fn find_nearest_enemy(ctx: &ReducerContext, position: DbVector2, player_id: u32) -> Option<DbVector2> {
    let mut nearest_enemy: Option<DbVector2> = None;
    let mut nearest_distance_squared = f32::MAX;

    // Iterate through all monsters in the game
    // TODO: can we use the spatial hash to speed this up?
    for boid in ctx.db.monsters_boid().iter() {
        // Calculate squared distance (more efficient than using square root)
        let dx = boid.position.x - position.x;
        let dy = boid.position.y - position.y;
        let distance_squared = dx * dx + dy * dy;

        // Check if this monster is closer than the current nearest
        if distance_squared < nearest_distance_squared {
            nearest_distance_squared = distance_squared;
            nearest_enemy = Some(boid.position);
        }
    }

    nearest_enemy
}

// Find the nearest enemy to a given position and return its entity ID
pub fn find_nearest_enemy_entity_id(ctx: &ReducerContext, position: DbVector2, _player_id: u32) -> Option<u32> {
    let mut nearest_enemy_id: Option<u32> = None;
    let mut min_dist_sq = f32::MAX;

    // Iterate over all monsters in the boid cache
    for boid in ctx.db.monsters_boid().iter() {
        // Skip monsters that are owned by the player (removed owner_id check since field doesn't exist)
        
        let dist_sq = (boid.position.x - position.x).powi(2) + (boid.position.y - position.y).powi(2);
        if dist_sq < min_dist_sq {
            min_dist_sq = dist_sq;
            nearest_enemy_id = Some(boid.monster_id);
        }
    }
    
    // Also check non-boid monsters
    for monster in ctx.db.monsters().iter() {
        // Skip monsters that are owned by the player (removed owner_id check since field doesn't exist)
        
        // Get monster position from boid data
        if let Some(boid) = ctx.db.monsters_boid().monster_id().find(&monster.monster_id) {
            let dist_sq = (boid.position.x - position.x).powi(2) + (boid.position.y - position.y).powi(2);
            if dist_sq < min_dist_sq {
                min_dist_sq = dist_sq;
                nearest_enemy_id = Some(monster.monster_id);
            }
        }
    }

    nearest_enemy_id
}

// Find a random enemy within a certain radius
pub fn find_random_target_in_radius(ctx: &ReducerContext, attacker_position: DbVector2, _attacker_player_id: u32, max_radius: f32) -> Option<DbVector2> {
    let mut targets_in_range: Vec<DbVector2> = Vec::new();
    let max_radius_squared = max_radius * max_radius;

    // Check all monsters within radius
    for boid in ctx.db.monsters_boid().iter() {
        let dx = boid.position.x - attacker_position.x;
        let dy = boid.position.y - attacker_position.y;
        let distance_squared = dx * dx + dy * dy;

        if distance_squared <= max_radius_squared {
            targets_in_range.push(boid.position);
        }
    }

    // Choose a random target from the list
    let mut rng = ctx.rng();
    if targets_in_range.is_empty() {
        return None;
    }
    let random_index = rng.gen_range(0..targets_in_range.len());
    Some(targets_in_range[random_index])
}

// Helper function to get a normalized direction vector, handling the zero-vector case
pub fn get_normalized_direction(direction: DbVector2) -> DbVector2 {
    if direction.x == 0.0 && direction.y == 0.0 {
        // If direction is zero, default to a standard direction (e.g., right)
        DbVector2::new(1.0, 0.0)
    } else {
        direction.normalize()
    }
}